### ISSUE TITLE
Local variables in frame.f_locals are not protected

### ISSUE THREAD

**Issue by joonis** [dashingsoft/pyarmor#1055]
Local variables in `frame.f_locals` are not protected in Pyarmor 8.0.22.
See: https://github.com/dashingsoft/pyarmor/issues/11#issuecomment-379503406

**Comment by jondy** [comment]
I'll add this feature in 8.1

**Comment by joonis** [comment]
Does not work in 8.1.1

**Comment by jondy** [comment]
Check it in a hello work script, now some functions may not obfuscated by wrap mode because it has special byte code, for non-wrap mode, no clear frame locals.

**Comment by joonis** [comment]
Obfuscated script:
```
def foo(arg=None):
    bar = 1
    raise RuntimeError('some error')
```

Now calling the function `foo`:

```
import sys
import mymodule
try:
    fintech.foo()
except:
    print(sys.exc_info()[2].tb_next.tb_frame.f_locals)
```

Results in:

`{'arg': None, '_var_var_0': 1}`

Linux, Python 3.8

**Comment by jondy** [comment]
I know this case. `raise` in the end results in function `foo` is obfuscated by non-wrap mode.

This will be improved later after I have full test cases, then I'll change this case to wrap mode. No full test cases I'm afraid it may broken others.

I'll feedback once it has been improved.

**Comment by jondy** [comment]
Fix in 8.1.9. 

It need enable
```
pyarmor cfg wrap_mode=2
```
In future version, wrap_mode = 2 may be default, but now it's not.

**Comment by joonis** [comment]
In combination with `clear_frame_locals=1` it now works for the example above.

But it segfaults with the original script. I try to strip it down to a smaller piece of code...

**Comment by joonis** [comment]
Now the following script segfaults with `clear_frame_locals=1` (regardless of the value of `wrap_mode`):
```
class MyClass:

    def foo(self):
        return super(MyClass, self).foo()
```

**Comment by jondy** [comment]
Thanks, I'll check it.

**Comment by jondy** [comment]
Fixed in `8.2.dev1`

**Comment by joonis** [comment]
The segfault has been fixed. Unfortunately it does not work in the following case.

Obfuscated script _mymodule.py_:

```
class MyClass:

    def foo(self, a=1):
        b = 2
        self.bar()
```

From an external script execute:

```
import sys
from mymodule import MyClass

class SubClass(MyClass):

    def bar(self):
        frame = sys._getframe(1)
        print(frame.f_code.co_name)
        print(frame.f_locals)

cls = SubClass()
cls.foo()
```

Output is:

```
foo
{'self': <__main__.SubClass object at 0x7fcbf5ca7250>, 'a': 1, '_var_var_0': 2}
```

**Comment by jondy** [comment]
@joonis it's better to discuss this by email pyarmor@163.com

**Comment by jondy** [comment]
Fixed in 8.2.dev2

**Comment by joonis** [comment]
> Fix in 8.1.9.
> 
> It need enable
> 
> ```
> pyarmor cfg wrap_mode=2
> ```
> 
> In future version, wrap_mode = 2 may be default, but now it's not.

Is `wrap_mode=2` still required? It works also with `wrap_mode=1` ...

**Comment by jondy** [comment]
A few functions need `wrap_mode=2`, most of functions `wrap_mode=1` is OK.
